### ISSUE TITLE
router_consistency job period 

### ISSUE THREAD

**Issue by luo985** [Kong/kong#4872]
router_consistency=eventual    : the router will be rebuilt asynchronously via a recurring background job running every second inside of each worker.

hello,  i need  every 10 second to run this job,  please add the job period config to kong.conf

- Kong version (1.2.0)

**Comment by p0pr0ck5** [comment]
What we be the use case for this?

**Comment by luo985** [comment]
I've been testing recently.  
if continuing change routes，the traffic fall to 30%.
then i change router_consistency period (by change source code) to 10 second,  the traffic  is more  better then 1 second
so i want to custom the period

**Comment by bungle** [comment]
@luo985 thanks for reporting. It seems rather strange as it is just a version check and if version is changed it is a mutex check. But I looked the code and it starts 2x<number-of-workers> every second, one to check plugins and one to check router, not sure if it helps to combine these two (I think my original patches ran just one per worker), but it will make a bit more latency to update other. Sure, running timers is away from processing request. Not sure if this is related with this comment: https://github.com/Kong/kong/issues/4850#issuecomment-517137349

**Comment by luo985** [comment]
@bungle  
when i testing， i change both check plugins job and check router job period to 10 seconds.
i think rebuild_router is relatively expensive， it affect the proxy traffic. 
so i want to do rebuild_router more infrequent.

**Comment by wurenhai** [comment]
I have the same issue too. There is a test case:

> service: 5000+
> route: 5000+
> upstream: 10000+
> router_consistency=eventual
> Kong version (1.2.0)

When I update service route 10 times pre second, by using _/routes/{name or id}_, CPU utilization rate will upgrade from 25% to 75%, and traffic will drop from 17000 to 5000.

Then I change rebuild_router  job period to 5 seconds. it will be good.


So I think rebuild_router is really expensive .

**Comment by bungle** [comment]
@luo985 / @wurenhai,

Can you try one thing, e.g. this:
https://github.com/Kong/kong/blob/next/kong/runloop/handler.lua#L816-L840

```lua
      timer_every(1, function(premature)
        if premature then
          return
        end

        -- Don't wait for the semaphore (timeout = 0) when updating via the
        -- timer.
        -- If the semaphore is locked, that means that the rebuild is
        -- already ongoing.
        local ok, err = rebuild_router(ROUTER_ASYNC_OPTS)
        if not ok then
          log(ERR, "could not rebuild router via timer: ", err)
        end
      end)

      timer_every(1, function(premature)
        if premature then
          return
        end

        local ok, err = rebuild_plugins_iterator(PLUGINS_ITERATOR_ASYNC_OPTS)
        if not ok then
          log(ERR, "could not rebuild plugins iterator via timer: ", err)
        end
      end)
```

to this:

```lua
      timer_every(1, function(premature)
        if premature then
          return
        end

        -- Don't wait for the semaphore (timeout = 0) when updating via the
        -- timer.
        -- If the semaphore is locked, that means that the rebuild is
        -- already ongoing.
        local ok, err = rebuild_router(ROUTER_ASYNC_OPTS)
        if not ok then
          log(ERR, "could not rebuild router via timer: ", err)
        end

        local ok, err = rebuild_plugins_iterator(PLUGINS_ITERATOR_ASYNC_OPTS)
        if not ok then
          log(ERR, "could not rebuild plugins iterator via timer: ", err)
        end
      end)
```

Aka only start one timer per worker per second. Does it have any impact? Probably not, but just want to check if that could help too. Yes, router rebuilding could be pretty expensive.

**Comment by luo985** [comment]
@bungle 
hello, we make a test as you asked.   it's not helpful. 
so i think increase period of the timer might be a good idea  (custom the period in kong.conf).

**Comment by bungle** [comment]
@luo985, thank you! And in someways I feel good that it didn't make the difference. Yes we can consider it to be configurable. Thanks.

**Comment by bungle** [comment]
@luo985 there is now a PR: #4897. Mind you test it out?

**Comment by bungle** [comment]
Closing this as the requested feature was implemented (currently going to `1.4.0`, but available in our `next` branch).
